### DISCUSSION TITLE
How to make keyPath for @Persisted property?

### DISCUSSION THREAD

**Discussion by yadabygit** [realm/realm-cocoa#8139] [Q&A]
i update Realm version to 10.35.1 in app.

Previously, we used `NSPredicate` for filtering and `NSSortDescriptor` for sorting objects. Both `NSPredicate` and `NSSortDescriptor` used structures like this `#keyPath(Entity.id)` to define the field to filter or sort. In general, it looks like this:
`NSSortDescriptor(key: #keyPath(Entity.id), ascending: true)` or `NSPredicate(format: "%K == %@", #keyPath(Entity.id), id)`.

I know that it is possible to use the field name directly. For example: `NSSortDescriptor(key: "id", ascending: true)`. But our approach is used because the application is large and we would not like to look for errors at runtime if we forget to update some key after changing the name of the object field.

Can you please tell me if it is still possible to use #keyPath for @Persisted properties or maybe there is another approach for sorting?

`NSSortDescriptor(keyPath \Entity.id, ascending: true)` initiator crashes with an error

![Screenshot 2023-02-17 at 12 28 06](https://user-images.githubusercontent.com/41676569/219606008-78ac4d65-9d95-4108-8d1a-e17a71cc908c.png)

**Comment by yadabygit** [comment]
RealmSwift.SortDescriptor works fine

**Comment by ari-berisha** [comment]
I ran into this problem when updated realm object models to @Persisted properties what I did is this extension 
```
extension KeyPath {
    var propertyPath: String {
        let fullPath = String(describing: self)
        
        guard let firstDot = fullPath.firstIndex(of: ".") else {
            return fullPath
        }
        
        let propertyOnly = fullPath[fullPath.index(after: firstDot)...]
        return propertyOnly.replacingOccurrences(of: "?", with: "")
    }
}
```
 
using this extension I can write the following "\((\Entity.id).propertyPath)" which translates to "id"

hope this helps.
